Global styles: enable user global styles revisions

ramonjd opened the pull request:
Related to:

- https://github.com/WordPress/gutenberg/issues/49601 (parent issue)
- https://github.com/WordPress/gutenberg/issues/41232
- https://github.com/WordPress/gutenberg/issues/33094

## What?

Display revisions to a user's global style changes in the global styles sidebar (revisions for the `wp_global_styles` custom post type).

These revisions are the ones made by the current, logged-in author. (should we show all authors' revisions?)

We're only showing maximum 100 to limit the payload.


https://user-images.githubusercontent.com/6458278/231377991-7b3b9905-104c-4857-9f94-b7de66a14c2f.mp4


## How?

By:
- adding a new `'version-history'` entry to `_links`, mimicking the one from the posts API that contains `count` and `href` properties
- adding a new callback to the global styles API – `/revisions`, which also mimics the REST API's `/wp/v2/posts/<parent>/revisions` – that returns data from `wp_get_post_revisions()`


## Testing Instructions


`npm run test:unit:php -- --filter Gutenberg_REST_Global_Styles_Controller_Test`

`npm run test:e2e:playwright -- --headed site-editor/user-global-styles-revisions.spec.js`

ramonjd committed:
Initial commit
Showing revisions for the wp_global_styles custom post type in the styles sidebar.
We're only showing max 10 to limit the payload.

ramonjd committed:
Updating test

ramonjd committed:
Sending and checking for settings as well since some style variations might have different settings we'd want to reinstate.
Doing a deep equality check to highlight the currently selected revision.

ramonjd committed:
Added a new endpoint to the global styles rest API (/revisions)
- Limit is 100 revisions in response
Added experimental revisions selector/action/resolver to the core-data store
Abstracted compare revisions into isGlobalStyleConfigEqual and added tests
Showing selector control in the UI if revision count is > 10

ramonjd committed:
revisions is no longer a property of the response object

tellthemachines commented on lib/compat/wordpress-6.2/class-gutenberg-rest-global-styles-controller-6-2.php:
```diff
@@ -16,9 +16,147 @@ class Gutenberg_REST_Global_Styles_Controller_6_2 extends WP_REST_Global_Styles_
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'args'                => array(
+						'id' => array(
+							'description'       => __( 'The id of a template' ),
```
Should this be "the id of a revision"?

ramonjd replied:
It definitely should be. 😄

ramonjd replied:
Actually no, now I've thought about it. It should be the id of a custom global styles post. Using that id we fetch the revisions.

tellthemachines commented on packages/edit-site/src/components/global-styles/screen-revisions.js:
```diff
@@ -0,0 +1,163 @@
+	);
+
+	const RevisionsComponent =
+		userRevisions.length >= 10 ? RevisionsSelect : RevisionsButtons;
```
We're not actually showing more than the 10 latest revisions, is that due to API response limits?

ramonjd replied:
Oh, I wanted to bump up the max number in the response to `100` but it turns out that I [forgot](https://github.com/WordPress/gutenberg/pull/46667/files#diff-265a1cbfa316459a657c81af12ad3f459bb2808f71758d11c163332bcb025742R129) 🤦

tellthemachines reviewed (commented):
> Is adding a revisions property to the global styles response a good idea? Should we rather add a new entry to _links and add a separate endpoint for revisions to avoid huge responses?

Are there any scenarios where we might want to fetch global styles (or part of it) _without_ the revisions? If so, yeah it's probably better to have a separate endpoint for them. 

Testing this a bit, it's working pretty well for under 10 revisions, but started acting funny once I added a bunch of them in a row (I didn't always reload the page between saves, not sure if that might have had any influence).

As mentioned below, only the last 10 revisions are appearing in the dropdown. 

I was switching between style variations in TT3, and ended up with all the variations showing gradient backgrounds:

<img width="275" alt="Screenshot 2022-12-22 at 5 28 42 pm" src="https://user-images.githubusercontent.com/8096000/209073891-da20c2b6-d2a1-4847-bbcc-62d6dae6ada9.png">
This happened when I tried to go back to a previous revision after switching to the sherbet variation and fiddling with a bunch of block styles.

Also, changes to block global styles weren't reverted when going back to a previous revision.

ramonjd commented:
> I was switching between style variations in TT3, and ended up with all the variations showing gradient backgrounds:

Good pick up, thank you! I thought I saw that too but was hoping that it might have been also happening on trunk. Turns out hope is a dangerous thing :)

ramonjd committed:
Update description for rest param
Bumped revision count that can be returned from the API to 100

ramonjd committed:
There's no reason to use the stored userConfig at all if we're reinstating the full revision.
The revision should overwrite everything.

ramonjd commented:
> started acting funny once I added a bunch of them in a row (I didn't always reload the page between saves, not sure if that might have had any influence).

> As mentioned below, only the last 10 revisions are appearing in the dropdown.

> I was switching between style variations in TT3, and ended up with all the variations showing gradient backgrounds

It actually is occurring on trunk. 

Steps to reproduce:

1. Fresh install of 2023 (no style customizations)
2. Site editor > activate Sherbet style variation. Save.
3. Change the background color by selecting a gradient from the default palette. Save. Refresh page.
4. See that the variations have all changed to that color.
5. No way back 🤷 


https://user-images.githubusercontent.com/6458278/209265668-3301620f-6ed5-413e-9195-cf93d5cefb6c.mp4


~~I'll create an issue.~~

Issue created: https://github.com/WordPress/gutenberg/issues/46758

ramonjd committed:
added a test for the revisions endpoint

ramonjd committed:
Added text domain

ramonjd committed:
Bumping minimum revision count until display to two:
- same behaviour as post editor
- since revisions includes the current saved post, so we need at least two in order to switch

ramonjd committed:
Added a very unpretty E2E test for global styles revisions. We can't yet delete global styles revisions and, because of other tests, we can't rely on the post revisions being a certain count so we try to cater for both in the test for now.
Added aria label to buttons in the UI

ramonjd committed:
Formatting the comments in the E2E test.
Checking for the number of revision buttons in the revisions list.

ramonjd commented on packages/core-data/src/resolvers.js:
```diff
@@ -453,7 +453,7 @@ __experimentalGetTemplateForLink.shouldInvalidate = ( action ) => {
 };
 
-export const __experimentalGetCurrentGlobalStylesId =
+export const __experimentalGetCurrentGlobalStyles =
```
The reason why I'm returning the entire global style object instead of just the ID is so we can access the [revisions url.](https://github.com/WordPress/gutenberg/pull/46667/files#diff-fec6390e24332dec4a0c7c74f9f7b9a015ce191e0e6348361ead793ceda3991dR509) 

E.g.: `currentGlobalStyles?._links?.[ 'version-history' ]?.[ 0 ]?.href;`

ramonjd replied:
It seemed a waste not to repurpose this method since it was fetching what we wanted anyway and only returning the id. 

Plus it's experimental so I didn't see much harm in changing it.

jameskoster commented:
Some questions from the design perspective:

* As already mentioned: should revisions be global or per user?
* Why does the button list switch to a dropdown, what are the conditions, is it necessary?
* Would a (minimally) visual treatment of each revision in the list be helpful? Similarly, should the list resemble a timeline (something like https://github.com/WordPress/gutenberg/discussions/37037).
* If you have unsaved changes, then select a revision, there doesn't appear to be a way back to those unsaved changes. Should there be a "Current (unsaved)" option in the list, or a warning that unsaved changes will be lost before switching to a revision?
* Is the access point appropriately prominent? Other options: the ellipsis menu, an icon next to the eye.
* Should the 'Reset to defaults' action be combined with revisions? There's some conceptual overlap there so perhaps its worth considering in the design.

cc @WordPress/gutenberg-design

ramonjd committed:
Fixing up things after a rebase, mainly usage of __experimentalGetCurrentGlobalStylesId, which has been removed in this PR.
Adding a reset button.

ramonjd committed:
Moving changes to global styles controller to 6.2 compat
Removing secondary button styling

ramonjd commented:
Thanks @jameskoster and sorry for the long delay between replies. This PR was a bit stale but I've managed to get it up and running again so that I can continue to improve (ignore failing tests for now please 😄)

> As already mentioned: should revisions be global or per user?

I'm not sure we can make that distinction in this case, since global styles and the associated controls, by their very nature, affect the site globally.

Do you mean that we should create a separate global styles post revision in the database for each user?

> Why does the button list switch to a dropdown, what are the conditions, is it necessary?

When there are more than 10 revisions it switches to a dropdown. The API will return up to 100 revisions.

I used the dropdown this way to avoid printing out 10++ buttons, but that was only a constraint of my implementation — if we change the UI then maybe we can think of a another way to preset many revisions.

> Would a (minimally) visual treatment of each revision in the list be helpful? Similarly, should the list resemble a timeline (something like https://github.com/WordPress/gutenberg/discussions/37037).

It might be. I'm thinking something akin to the way we preview block styles, with the overlay.

I think in the first iteration it'd be possible to skip this since clicking on a revision will load the styles immediately. The change only takes effect once the post is saved.

> If you have unsaved changes, then select a revision, there doesn't appear to be a way back to those unsaved changes. Should there be a "Current (unsaved)" option in the list, or a warning that unsaved changes will be lost before switching to a revision?

Great point 😄 Thanks for raising it. 

I'll have a look what's involved. I believe there's only one "state" in terms of global styles and it's the one we see in the editor. So we might have to create another temporary one to stash unsaved changes. 

Not sure. I'll have a think.

> Is the access point appropriately prominent? Other options: the ellipsis menu, an icon next to the eye.
Should the 'Reset to defaults' action be combined with revisions? There's some conceptual overlap there so perhaps its worth considering in the design.

Good questions.

I've added a pretty rudimentary reset button for now, just to play with functionality. 

As for the position of the revisions, I don't have a strong opinion right now. I chose the panel since it's in context and I could also leverage the existing navigation structure to create a sub page for the revisions list.

Near the eye icon lends sufficient context as well I imagine.

Mind you all this might change given that there's talk of exposing revisions functionality for templates, patterns etc. See https://github.com/WordPress/gutenberg/issues/33094 Perhaps we'll need a unified UI (?)

ramonjd committed:
JS LINT YO!

ramonjd committed:
Relocate js utils tests for isGlobalStyleConfigEqual

ramonjd committed:
Move rest route registration to compat 6.3

paaljoachim commented:
Testing
Gutenberg PR build
WordPress 6.2
Twenty Twenty Three

I did not see Styles -> Revisions show up when using the Gutenberg PR build...

ramonjd commented:
> I did not see Styles -> Revisions show up when using the Gutenberg PR build

Thanks a lot for testing @paaljoachim 

As far as I can see, the revisions do appear in the global styles sidebar (in edit mode) in this PR. 

There has to be at last one revision: 

![2023-04-11 14 55 35](https://user-images.githubusercontent.com/6458278/231060052-11589e77-450f-43d6-898e-27aa688877f6.gif)

Testing on http://gutenberg.run/46667

ramonjd committed:
update e2e tests so that they pass :D I was importing a non-existant method

ramonjd committed:
Testing with undo/reset buttons

ramonjd commented:
> If you have unsaved changes, then select a revision, there doesn't appear to be a way back to those unsaved changes. Should there be a "Current (unsaved)" option in the list, or a warning that unsaved changes will be lost before switching to a revision?

Fortunately the "undo" functionality works in this case. I've added a very fugly undo button in the latest commmit to test how it works.

paaljoachim commented:
Thank you Ramon!
I actually changed something in the Template and did not notice any revisions in the Styles area. But thanks to your above comment and gutenberg.run link I discovered the error I had made.

It makes me wonder if there should be an universal revisions for templates and styles when making an adjustment in the Site Editor. Or perhaps it is better to split these into two different revisions areas.

One in the Styles area and one Template tab area.

Back to Styles Revisions.

Clicking Revisions. 
![Screenshot 2023-04-11 at 12 54 03](https://user-images.githubusercontent.com/5323259/231139424-c972fcae-1c98-4ac0-9eda-d1a340e0bbb9.jpg)

**Undo icon**
The undo button points to the right.
The general Undo button in the Template and in a page/post points to the left like so:

![Screenshot 2023-04-11 at 12 56 19](https://user-images.githubusercontent.com/5323259/231139908-ef813f7b-e94d-4863-95f0-6c23d24faa82.jpg)

As the arrow that points to the right is the same icon used for redo in a page/post/template.

**Revisions**
We can see a list of revisions and can gradually move backward somewhat similar to an undo in a page/post/template.
One about taking it one step further. Click a single revision and have it open as an accordion to see the exact styles that were changed. Listing up the changes. Then a button below the list which takes one back. (I noticed James mentioned something similar.)

Revisions in general be it for posts/pages/template/styles etc all need an unified approach to how to handle this matter.

Thank you for working on revisions Ramon!

ramonjd committed:
Moving the revisions link so that it's under the drop down
Updating e2e tests
A sad attempt at caching the unsaved userConfig
Adding an `is_latest` prop to the rest response to indicate the latest saved revision

ramonjd commented:
> It makes me wonder if there should be an universal revisions for templates and styles when making an adjustment in the Site Editor. Or perhaps it is better to split these into two different revisions areas.

I think it would be great to have a visual diff on these things, and that includes template etc.

I just discovered that template have revisions since https://github.com/WordPress/gutenberg/pull/45215.

It uses the post diff viewer:

<img width="800" alt="Screenshot 2023-04-12 at 4 55 36 pm" src="https://user-images.githubusercontent.com/6458278/231376518-3bf08419-5591-436b-9e0b-26f46240945b.png">

I did think about using this view for the style config, but thought it would be a better experience to see the styles in the editor rather than a diff on JSON.

However I'm happy to follow that example for global styles if folks think it's better to have a unified experience. If I get time tomorrow/this week I'll throw a version that uses the diff tool so we can compare.

I've updated this PR to include a button to restore unsaved changes (and removed the janky undo).

<img width="283" alt="Screenshot 2023-04-12 at 4 58 53 pm" src="https://user-images.githubusercontent.com/6458278/231377044-5d67e183-9c07-41f0-a307-369aea5cd415.png">

The longer route would be to create a visual diff tool and unify these things as you suggest.

ramonjd committed:
Yoda?

jasmussen commented:
Revisions are such an understated feature of WordPress, the fact that everything has it and it works so well. We need these to be beautiful and excellent, have a revisions UI that is easily recognizable across all properties and excellent to use.

A bit longer term we also have an opportunity to be smarter about the diffs. Just like how GitHub provides a visual diff for images that are recognized, WordPress recognizes blocks, so outside of malformed block markup, we can show much more legible diffs by omitting the HTML delimiters, perhaps showing the diff in less of a code font, making it all the more ergonomic to review.

Thanks for working on it!

paaljoachim commented:
Btw I came across creating and editing custom block style variations in Global Styles.
https://github.com/WordPress/gutenberg/issues/49602#issuecomment-1505054467
Which made me think of this PR.

In having a consistent method in clicking a revision, style or any other item to view the smaller pieces of what it is made up of. As if it was somekind of List view to where one can for instance see a Group block and all the blocks it contains. 

**Inspiration**

One method is to be able to easily view what changes are made in a specific revision (style or other places) perhaps something similar to what James shared here: https://github.com/WordPress/gutenberg/discussions/37037

Another method might use a bigger modul to compare revisions and becomes a much bigger and complex project.
https://github.com/WordPress/gutenberg/issues/25270 and https://github.com/WordPress/gutenberg/issues/16053#issuecomment-673162147

Initially it seems like what was shared by James is a nice way forward which could perhaps also be used in other places in the Gutenberg UX.

jasmussen commented:
For reference, here is a draft and very much work in progress revisions panel I've been tinkering with for template part revisions, in case it might inspire further work:
<img width="1600" alt="Revisions in template inspector, inline panel" src="https://user-images.githubusercontent.com/1204802/231445604-b7c7c0e7-2913-454f-b983-c4c3dbe7cb2d.png">

ramonjd commented:
Thanks for the input folks. It's really helpful.

> A bit longer term we also have an opportunity to be smarter about the diffs. Just like how GitHub provides a visual diff for images that are recognized, WordPress recognizes blocks, so outside of malformed block markup, we can show much more legible diffs by omitting the HTML delimiters, perhaps showing the diff in less of a code font, making it all the more ergonomic to review.

> In having a consistent method in clicking a revision, style or any other item to view the smaller pieces of what it is made up of. As if it was somekind of List view to where one can for instance see a Group block and all the blocks it contains.

100% 

If we can render block examples, and also the style book view in the site editor, then a visual diff is definitely within our grasp.

 I also like the parallels drawn to the style variation example.

> Revisions are such an understated feature of WordPress, the fact that everything has it and it works so well. We need these to be beautiful and excellent, have a revisions UI that is easily recognizable across all properties and excellent to use.

Great point.

I'm hesitant to go too fast with this PR then. I definitely do not want to introduce a UI/UX in a hurry, and would be keen to iterate alongside @jasmussen's design explorations 👍 

I'm thinking of throwing up a PR that would introduce the existing revisions UI for global styles – similar to how [templates now](https://github.com/WordPress/gutenberg/pull/45215) (and all posts) work, that is, using the existing code diff checker.

That would allow styles revisions, simple and cumbersome, but at least we'd have some of the framework ready on which to build a more considered design. 

What do folks think?

ramonjd committed:
Adding author display name and avatar to rest response
Redesigning the buttons view
Rationalizing rest API return properties

ramonjd commented:
> I'm thinking of throwing up a PR that would introduce the existing revisions UI for global styles – similar to how https://github.com/WordPress/gutenberg/pull/45215 (and all posts) work, that is, using the existing code diff checker.

Okay, I tried this and there are some documented bugs with this approach:

1. The first revision is empty: https://github.com/WordPress/gutenberg/issues/49164
2. The "← Go to editor" doesn't work: https://github.com/WordPress/gutenberg/issues/48065

https://user-images.githubusercontent.com/6458278/231662130-115e016b-0d00-4bf7-8352-cf36f799a270.mp4

jasmussen commented:
> I'm hesitant to go too fast with this PR then

Perhaps on the contrary, I'm excited to see momentum on this exactly because it's such an understated feature. My personal take, but happy to hear other opinions (@WordPress/gutenberg-design), I'd love to see an interface rolled out sooner rather than later. So long as it is componentized and re-used across the various places, so it's a single piece of software we can iterate, I'd rather surface access to it sooner rather than later.

paaljoachim commented:
I very much agree on Joen's above comment.

There is a very good momentum in continuing the progress, bugs show up that needs to be fixed and along the way the general revision system we have today will be very much improved.

We are here to also test and give feedback.

Thank you for working on the revision system Ramon!

annezazu commented:
Agreed! We can wrap this into a call for testing to get feedback along the way too (not this next one coming up this week but the _next_ one after that in a few weeks).

ramonjd commented:
Thanks for the encouragement folks. I'll keep at it 😄

ramonjd commented:
I updated the design of the revisions list (thanks for the design images @jasmussen!)

<img width="299" alt="Screenshot 2023-04-14 at 4 10 16 pm" src="https://user-images.githubusercontent.com/6458278/231958501-336b211b-f74d-453f-b27b-b3bddafc8c88.png">

I ran out of time, but I also wanted to test an idea I had for a visual preview panel for revisions, structurally similar to the style book which stands visually and contextually independent of the template being edited.  

As it stands, selecting a revision in this PR loads that revision in the editor. This is great for previewing but brings about the need for the "restore unsaved changes" button, whose logic is a bit convoluted.

A visual panel could be used for templates, and maybe even have a side-by-side view. I have something in mind, so I'll try to sketch it out/design it next week. Words are failing me today. Thank you!

jasmussen commented:
Nice work.

> As it stands, selecting a revision in this PR loads that revision in the editor. This is great for previewing but brings about the need for the "restore unsaved changes" button, whose logic is a bit convoluted.

> A visual panel could be used for templates, and maybe even have a side-by-side view. I have something in mind, so I'll try to sketch it out/design it next week. Words are failing me today. Thank you!

I like how Google Docs does revisions. Which is a bit of an undertaking. Depending on the challenge, it seems okay to take you out of the editor to the legacy screens in the interim.

ramonjd committed:
Update PHP tests title > date

ramonjd commented:
I've been running around in circles in an attempt to come up with a reasonable workflow today, and have concerns about how to deal with unsaved changes in the editor.

I'm worried that users won't know that loading a revision overwrites the styles in the editor canvas, thereby removing any unsaved changes. For example, a user might want to browse previous revisions only. 

@jameskoster [raised this concern](https://github.com/WordPress/gutenberg/pull/46667#issuecomment-1370903370) at the start of this PR, and suggested that a warning might be temporary approach until it's possible to load revision styles independent to the styles that show in the editor.

This is what I've come up with following those lines:

https://user-images.githubusercontent.com/6458278/232691886-bffc6043-b9cf-4e9b-8f42-2032f20d72c3.mp4

I haven't committed it to this PR yet. 

Thinking more broadly, I have positive feelings towards an in-editor revisions view, let's say [something like the style book](https://github.com/WordPress/gutenberg/issues/49601#issuecomment-1512357036) where we could compile and load any style without affecting the main editor canvas.

That would require a major refactor, but it might be useful to serve as a basis for comparing template revisions as well. I'd like to explore this option in hack branch.

jameskoster commented:
> I'm worried that users won't know that loading a revision overwrites the styles in the editor canvas, thereby removing any unsaved changes. For example, a user might want to browse previous revisions only.

Yeah this feels like a blocker to me. I know we have the Undo affordance, but it doesn't feel intuitive in this case. 

What you shared above seems like a good start. The action(s) are significant enough that it warrants being placed in a modal imo. We could potentially invoke a similar UI in other flows where customisations are lost such as switching variations: https://github.com/WordPress/gutenberg/issues/40561

ramonjd commented:
Thanks folks. @andrewserong and @tellthemachines had some good ideas that I can trial as well: namely, auto-saving unsaved changes (if any) when opening the revisions panel. This would technically create a new revision I think, but I'm interested to see how it would work in practice. 

> The action(s) are significant enough that it warrants being placed in a modal imo. We could potentially invoke a similar UI in other flows where customisations are lost such as switching variations: https://github.com/WordPress/gutenberg/issues/40561

This makes perfect sense too. I'll play around.

andrewserong commented:
> Thanks folks. @andrewserong and @tellthemachines had some good ideas

It was all @tellthemachines' idea, I was just being the cheer squad 😄

Thanks for digging into this feature so deeply!

ramonjd commented:
> What you shared above seems like a good start. The action(s) are significant enough that it warrants being placed in a modal imo. We could potentially invoke a similar UI in other flows where customisations are lost such as switching variations:

My plan is to implement this in this PR, but [I took a stab at repurposing the stylebook code to load blocks and styles independently](https://github.com/WordPress/gutenberg/issues/49601#issuecomment-1514183504).  The main benefit is that users can open revisions, try them out and return to the editor in whatever state they left it in, that is, any unsaved changes are unaffected.

There's still the need for a prompt to save unsaved changes before a revision is restored just in case.

ramonjd commented:
Closing this PR and tracking follow ups based on the [experimental PR](https://github.com/WordPress/gutenberg/pull/49912) over in the issue:
- https://github.com/WordPress/gutenberg/issues/49601
